src/session/mpi.rs

```diff
@@ -252,25 +256,17 @@ impl SessionMPI {
         let app_path = "app".to_owned();
         let tarball_path = config_path.join(&sources.path);
 
-        let tarball = fs::read(tarball_path).map(Into::into);
-        let tarball_stream = futures::stream::once(tarball);
-
         let deployments: Vec<_> = self
             .providers
             .iter()
             .map(|provider| provider.session.clone())
             .collect();
 
-        self.hub_session
-            .new_blob()
-            .from_err()
-            .and_then(move |blob| {
-                blob.upload_from_stream(tarball_stream)
-                    .from_err()
-                    .and_then(move |_| Ok(blob))
-            })
+        self.upload_to_hub(&tarball_path)
+            .context("uploading the source tarball")
```

Comment:
Is `context` coming from your `failure-ext` crate?

Reply:
Yes and no. 
No, because this branch doesn't depend on the `failure_ext` crate yet.
Yes, because this is `FutureExt::context`, which was moved from `src/future_ext.rs` to a separate crate.